Mathlib/Algebra/AddConstMap/Equiv.lean

```diff
@@ -0,0 +1,108 @@
+/-
+Copyright (c) 2024 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
+-/
+import Mathlib.Algebra.AddConstMap.Basic
+/-!
+# Equivalences conjugating `(· + a)` to `(· + b)`
+
+In this file we define `AddConstEquiv G H a b` (notation: `G ≃+c[a, b] H`)
+to be the type of equivalences such that `∀ x, f (x + a) = f x + b`.
+
+We also define the corresponding typeclass and prove some basic properties.
+-/
+
+open Function
+open scoped AddConstMap
+
+/-- An equivalence between `G` and `H` conjugating `(· + a)` to `(· + b)`. -/
+structure AddConstEquiv (G H : Type*) [Add G] [Add H] (a : G) (b : H)
+  extends G ≃ H, G →+c[a, b] H
+
+/-- Interpret an `AddConstEquiv` as an `Equiv`. -/
+add_decl_doc AddConstEquiv.toEquiv
+
+/-- Interpret an `AddConstEquiv` as an `AddConstMap. -/
+add_decl_doc AddConstEquiv.toAddConstMap
+
+@[inherit_doc]
+scoped [AddConstMap] notation:25 G " ≃+c[" a ", " b "] " H => AddConstEquiv G H a b
+
+/-- A typeclass saying that `F` is a type of bundled equivalences `G ≃ H`
+semiconjugating `(· + a)` to `(· + b)`. -/
+class AddConstEquivClass (F : Type*) (G H : outParam (Type*)) [Add G] [Add H]
+    (a : outParam G) (b : outParam H) extends EquivLike F G H where
+  /-- A map of `AddConstEquivClass` class semiconjugates shift by `a` to the shift by `b`:
+  `∀ x, f (x + a) = f x + b`. -/
+  map_add_const (f : F) (x : G) : f (x + a) = f x + b
+
+instance (priority := 100) {F : Type*} {G H : outParam Type*} [Add G] [Add H]
+    {a : outParam G} {b : outParam H} [h : AddConstEquivClass F G H a b] :
+    AddConstMapClass F G H a b where
+  toDFunLike := inferInstance
+  __ := h
+
+instance {G H : Type*} [Add G] [Add H] {a : G} {b : H} :
+    AddConstEquivClass (G ≃+c[a, b] H) G H a b where
+  coe f := f.toEquiv
+  inv f := f.toEquiv.symm
+  left_inv f := f.left_inv
+  right_inv f := f.right_inv
+  coe_injective' | ⟨_, _⟩, ⟨_, _⟩, h, _ => by congr; exact DFunLike.ext' h
+  map_add_const f x := f.map_add_const' x
+
+namespace AddConstEquiv
+
+variable {G H K : Type*} [Add G] [Add H] [Add K] {a : G} {b : H} {c : K}
+
+@[ext] lemma ext {e₁ e₂ : G ≃+c[a, b] H} (h : ∀ x, e₁ x = e₂ x) : e₁ = e₂ := DFunLike.ext _ _ h
+
+lemma toEquiv_injective : Injective (toEquiv : (G ≃+c[a, b] H) → G ≃ H) :=
```

Comment:
should this not be done by some general `EquivLike` plumbing?

Reply:
I see that other `*Equiv` prove `toEquiv_injective` first, then use it for `EquivLike`. I changed this file to follow the same pattern. We may change it everywhere later.